### PR TITLE
two android bugfixes: blankscreen and blackscreen + reduce apk size

### PR DESCRIPTION
**Pull request by cerisara**
Two bugfixes and one enhancement for android:
- fix blankscreen when launching app on some androids
- fix blankscreen when pushing app in the background and back in the foreground on some androids
- fix blackscreen when pressing the "back button" and then calling back the app
- fix blackscreen when rotating the screen

- reduce the size of the apk with the option "android/arm"

### CODE REVIEW

**Comment by andydotxyz** [comment]
Testing on a Galaxy S10 with this patch enabled on rotation I get a blank screen and:
Fyne error:  OpenGL Error: 1281
02-17 18:27:05.324 10432 10483 I GoLog   :   At: /Users/andy/Code/Go/src/github.com/fyne-io/examples/vendor/fyne.io/fyne/internal/driver/gomobile/driver.go:146

**Comment by andydotxyz** [comment]
Looking into it I wonder, should this commit be reverted alongside your changes to work (relates to the terminate issue) - it feels like it might not have been the right fix? https://github.com/fyne-io/mobile/commit/5c530ec02bccf4915d45b9c5915bf35f18b84831#diff-dc72d6242b8bdb49da90dd4f078b2282

**Comment by cerisara** [comment]
> Looking into it I wonder, should this commit be reverted alongside your changes to work (relates to the terminate issue) - it feels like it might not have been the right fix? [fyne-io/mobile@5c530ec#diff-dc72d6242b8bdb49da90dd4f078b2282](https://github.com/fyne-io/mobile/commit/5c530ec02bccf4915d45b9c5915bf35f18b84831#diff-dc72d6242b8bdb49da90dd4f078b2282)

Yes, looks like this 2018 fix may be part of the cause of today's problem... But to be honest, I'm not good enough in open-gl to even understand what was this previous commit fixing... Well I partly understand it was about keeping the surface safe for later reuse, but the choice to quit on a dead signal may not have been the best.

I've traced a vanilla gomobile app, and it was not quitting the main loop on the back button. So it looks like it's the standard way for android apps to stay loaded in RAM, at least with gomobile. I've checked also with one of my java app, and it also stays in RAM until the user press the "RAM clean" android widget. So I don't see any difference between go apps (with my patch) and java apps in terms of RAM lifecycle ?

File: vendor/golang.org/x/mobile/app/android.go [hunk 1]
```diff
@@ -363,7 +363,7 @@ func mainUI(vm, jniEnv, ctx uintptr) error {
 			C.surface = nil
 			theApp.sendLifecycle(lifecycle.StageAlive)
 		case <-activityDestroyed:
-			C.terminate()
+			// C.terminate()
```
[the remark this answers is not in the document]
Comment (cerisara) [reply vendor/golang.org/x/mobile/app/android.go]: I'm really not sure it should ever quit this loop, except after a RAM clean command by the user. But I may be wrong. If we want to really quit this loop, then we also must manage a way to resume the app cleanly on signal.

File: internal/driver/gomobile/driver.go [hunk 2]
```diff
@@ -129,6 +134,11 @@ func (d *mobileDriver) Run() {
 			case size.Event:
 				currentSize = e
 			case paint.Event:
+				// this is a fix to prevent the window to appear as a "blank screen" on android
+				// on some android phones, we may get a 0x0 size
+				if currentSize.WidthPx == 0 {
```
[the remark this answers is not in the document]
Comment (cerisara) [reply internal/driver/gomobile/driver.go]: Yes you're right; I edit that

File: vendor/golang.org/x/mobile/app/android.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply vendor/golang.org/x/mobile/app/android.go]: If someone calls app.Quit() we really should quit the loop. It's not a common use-case but could also be used to tidy up error cases cleanly which is just good housekeeping

**Comment by andydotxyz** [comment]
I think you might be right.
The patch was introduced to reset the GL context so it would be re-created later I think, however this seems to not match the fyne app lifecycle which is probably the cause of the problem. If you think that, plus this patch, solves things I can revert that change on our mobile fork so you can reference a new version in go.mod.

File: vendor/golang.org/x/mobile/app/android.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cerisara) [reply vendor/golang.org/x/mobile/app/android.go]: I'm not so sure about that in android: if you look at https://developer.android.com/guide/components/activities/process-lifecycle they say "will remain running until it is no longer needed and the system needs to reclaim its memory" and "An unusual and fundamental feature of Android is that an application process's lifetime is not directly controlled by the application itself".

In fyne, in the type App interface, it's also said for the method "Quit": "This should be used with caution, many platforms discourage exiting an application from within the code and some mobile systems do not allow it."

The problem may be that if the app quits without being killed by Android, then the Android system may keep a "background" image of the app, and may try to restore a dead process, leading to the black screen.

But I'm really not an expert, so it's OK for me whatever you think should be done.

**Comment by cerisara** [comment]
> If you think that, plus this patch, solves things I can revert that change on our mobile fork so you can reference a new version in go.mod.

I'm not so sure to understand what you want to do: you want to revert the 2018 patch completely (but could it be useful for something else ?) ? And further apply my patch (which afaik is partly equivalent to reverting the 2018 patch) ? That could be an option, but we'll need to test it afterwards anyway. The other options are either just reverting the 2018 patch, or just applying my patch, I don't know; may be doing both is safer indeed... we'll need to test anyway.

File: vendor/golang.org/x/mobile/app/android.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply vendor/golang.org/x/mobile/app/android.go]: Yes, after reading more I think you are right... basically the "right fix" would probably to be revert the change that resolved this bug? https://github.com/fyne-io/fyne/issues/580

File: vendor/golang.org/x/mobile/app/app.go [hunk 3]
```diff
@@ -140,7 +140,6 @@ func (a *app) HideVirtualKeyboard() {
 	driverHideVirtualKeyboard()
 }
 
-
```
Comment (andydotxyz) [inline vendor/golang.org/x/mobile/app/app.go]: p.s. best not to commit format changes in vendor files - they will just get overwritten by the next vendor update so it's just noise

**Comment by andydotxyz** [comment]
> > If you think that, plus this patch, solves things I can revert that change on our mobile fork so you can reference a new version in go.mod.
> 
> I'm not so sure to understand what you want to do: you want to revert the 2018 patch completely (but could it be useful for something else ?) ? And further apply my patch (which afaik is partly equivalent to reverting the 2018 patch) ? That could be an option, but we'll need to test it afterwards anyway. The other options are either just reverting the 2018 patch, or just applying my patch, I don't know; may be doing both is safer indeed... we'll need to test anyway.

I have updated our mobile fork to revert terminate() and tested it locally. I kept in part of the patch so we'll see how it goes. If you pull in the latest develop (55d04ef) then you should see that in the vendor folder, then your patch without modifying terminate() should work as before.

File: internal/driver/gomobile/driver.go [hunk 4]
```diff
@@ -167,7 +177,8 @@ func (d *mobileDriver) Run() {
 			}
 
 			if d.quit {
-				break
+				// bugfix ? when pressing the back button, other gomobile apps do not terminate, so why should we ?
```
[the remark this answers is not in the document]
Comment (andydotxyz) [reply internal/driver/gomobile/driver.go]: After reading more I think you are right... basically the "right fix" would probably to be revert the change that resolved this bug? #580

**Comment by andydotxyz** [comment]
After further testing I don't think that this fixes the case:
* fix blackscreen when pressing the "back button" and then calling back the app

But happy to land the fixes for all other items and work on that separately

**Comment by andydotxyz** [comment]
I have just added another commit on develop that adds device scaling according to DPI.
This may conflict with this patch slightly, apologies for that. I thought that it would add to rounding out the older device support, so hopefully it works well with your changes.

**Comment by andydotxyz** [comment]
I'd like to start work on polishing a 1.2.3 release shortly and ideally we would get these Android fixes in - do you think this would be possible @cerisara?

**Comment by cerisara** [comment]
Hi Andy,

Yes, sorry for my lack of reaction, but my work load has increased a lot in the last two days, and
I pitifully don't have much time to go on testing these fixes for now. :-(

The status of my last tests were:

- I don't know the impact of reversing the 2018 commit; I've not had time to test it
- I don't know how to move the fix out of vendor/ and where to put it instead: I've not had time to think about it
- All that I'm sure of (because I've made extensive tests on my Android 7.0 phone) is that the commit 1578e070a4c1b in here: https://github.com/cerisara/fyne/tree/blackscreen is working fine, including moving away from the app/pressing the back button/rotating/ and coming back to the app

Sorry I don't have fresher news, but I'm really very busy at the moment...


----- Mail original -----

**Comment by andydotxyz** [comment]
Thanks for the info.

I have reverted the 2018 fix and tested with your changes - that is what got me to the place where it all seemed to work apart from the back button.
The vendor updates are all done on develop already, along with the DPI change I mentioned. If you pull in the latest version of the branch you should just get it all :)

**Comment by cerisara** [comment]
OK, I've made 2 tests:

1- I've pulled your develop branch, the app is launching fine on my phone, but the rest fails (rotation, back button, background/restore)
2- I've merged your develop branch into my develop branch, and nothing works, not even launching the app

So far, for me, only my blackscreen branch is solving all the issues afaik. Something's still not right I'm afraid...

**Comment by andydotxyz** [comment]
Oh goodness that is no good at all!
Are you testing using one of the fyne apps? (I.e. fyne_demo) id like to rule out module or vendor issues which can confuse things

**Comment by cerisara** [comment]
Sorry, but I'm not sure any more that this PR will solve the android bugs; so it's best to close it.
I'm only confident in my blackscreen branch, but it's too far away from the develop HEAD, so we'll need to reinject each part of the fix in the latest dev and test on real devices: such careful real device testing takes a lot of time (I've already gone through it once) and I really can't afford it for now, sorry about that. A better option would be to automate the test for this bug, but I've no idea how.

**Comment by andydotxyz** [comment]
Ok, sorry if the changes on develop made this harder. I will try to work on merging patches from https://github.com/cerisara/fyne/tree/blackscreen to see if I can get something for 1.2.3 in the next few days.

**Comment by andydotxyz** [comment]
The patches made it to release/v1.2.3 for testing

**Comment by andydotxyz** [comment]
V1.2.3 is now out, please let us know how you get on.
Unfortunately the APK size commit did not make it in - It would be great if you could open a new PR with that, but also expanded to allow any valid arch as the second part of the parameter and include it in the command docs so others can discover it.
Many thanks in advance!

**Pull request closed without merging**